src/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -282,262 +280,530 @@ internal static ulong ToUInt64(object value)
-            return retValue;
+            ReadOnlySpan<char> valueSpan = value.AsSpan().TrimStart();
+            if (valueSpan.Length == 0)
+            {
+                throw new ArgumentException(SR.Arg_MustContainEnumInfo, nameof(value));
+            }
+
+            bool success;
+            TEnum result;
+            
+            switch (Type.GetTypeCode(typeof(TEnum)))
+            {
+                case TypeCode.SByte: success = TryParseSByteEnum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+                case TypeCode.Byte: success = TryParseByteEnum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+                case TypeCode.Int16: success = TryParseInt16Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+                case TypeCode.UInt16: success = TryParseUInt16Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+                case TypeCode.Int32: success = TryParseInt32Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+                case TypeCode.UInt32: success = TryParseUInt32Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+                case TypeCode.Int64: success = TryParseInt64Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+                case TypeCode.UInt64: success = TryParseUInt64Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+                default: success = TryParseRareEnum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: true, out result); break;
+            }
+
+            Debug.Assert(success);
+            return result;
         }
 
-        public static bool TryParse<TEnum>(string value, out TEnum result) where TEnum : struct
+        private static bool TryParseSByteEnum<TResult>(RuntimeType enumType, string originalValueString, ReadOnlySpan<char> value, bool ignoreCase, bool throwOnFailure, out TResult enumResult)
         {
-            return TryParse(value, false, out result);
+            Debug.Assert(typeof(TResult) == enumType || typeof(TResult) == typeof(object));
+            Debug.Assert(enumType.GetEnumUnderlyingType() == typeof(sbyte));
+
+            bool failureIsOverflow = false;
+            if (StartsNumber(value[0]) && Number.TryParseInt32IntegerStyle(value, NumberStyles.AllowLeadingSign | NumberStyles.AllowTrailingWhite, CultureInfo.InvariantCulture.NumberFormat, out int intResult, ref failureIsOverflow))
+            {
+                if ((uint)(intResult - sbyte.MinValue) <= (sbyte.MaxValue - sbyte.MinValue))
```

Comment:
Pass in the range as an argument and have one method to parse all SByte/Short/Int32 to avoid code duplication?

Reply:
Yup.  Same for UInt32/UInt16/Byte.